src/libraries/Common/tests/TestUtilities/System/Buffers/BoundedMemory.Unix.cs

```diff
@@ -1,45 +1,270 @@
+using System.Runtime.InteropServices;
+using System.Security;
+
+using Xunit;
+
 namespace System.Buffers
 {
     public static partial class BoundedMemory
     {
-        private static UnixImplementation<T> AllocateWithoutDataPopulationUnix<T>(int elementCount, PoisonPagePlacement placement) where T : unmanaged
+        private unsafe static UnixImplementation<T> AllocateWithoutDataPopulationUnix<T>(int elementCount, PoisonPagePlacement placement) where T : unmanaged
         {
-            // On non-Windows platforms, we don't yet have support for changing the permissions of individual pages.
-            return new UnixImplementation<T>(elementCount);
+            long cb, totalBytesToAllocate;
+            checked
+            {
+                cb = elementCount * sizeof(T);
+                totalBytesToAllocate = cb;
+
+                // We only need to round the count up if it's not an exact multiple
+                // of the system page size.
+
+                long leftoverBytes = totalBytesToAllocate % SystemPageSize;
+                if (leftoverBytes != 0)
+                {
+                    totalBytesToAllocate += SystemPageSize - leftoverBytes;
+                }
+
+                // Finally, account for the poison pages at the front and back.
+
+                totalBytesToAllocate += 2 * SystemPageSize;
+            }
+
+            // Reserve and commit the entire range as NOACCESS.
+
+            void* ptr = UnsafeNativeMethods.mmap64(
+                addr: null,
+                length: (nuint)totalBytesToAllocate /* cast throws OverflowException if out of range */,
+                prot: MemoryMappedProtections.PROT_NONE,
+                flags: MemoryMappedFlags.MAP_PRIVATE | MemoryMappedFlags.MAP_ANONYMOUS,
+                fd: -1,
+                offset: 0);
+
+            if (ptr == null || ptr == (void*)UIntPtr.MaxValue)
+            {
+                throw new Win32Exception();
+            }
+
+            // Done allocating! Now carve out a READWRITE section bookended by the NOACCESS
+            // pages and return that carved-out section to the caller. Since memory protection
+            // flags only apply at page-level granularity, we need to "left-align" or "right-
+            // align" the section we carve out so that it's guaranteed adjacent to one of
+            // the NOACCESS bookend pages.
+            MemoryMappedHandle handle = new((IntPtr)ptr, (nuint)totalBytesToAllocate);
+            return new UnixImplementation<T>(
+                handle,
+                byteOffsetIntoHandle: (placement == PoisonPagePlacement.Before)
+                    ? SystemPageSize /* just after leading poison page */
+                    : checked((int)(totalBytesToAllocate - SystemPageSize - cb)) /* just before trailing poison page */,
+                elementCount: elementCount)
+            {
+                Protection = MemoryMappedProtections.PROT_READ | MemoryMappedProtections.PROT_WRITE
+            };
```

Comment:
Why can't this just use MemoryMappedFile.CreateNew rather than direct use of mmap?

Reply:
1. I didn't realise that thing existed (😅)
2. I'm just not sure how portable that would be - we'll be basically relying on MemoryMappedFile using some function that works with `mprotect`, and the only thing I know about it is that only mappings that have been created through `mmap` are 'guaranteed' to work ("POSIX says that the behavior of mprotect() is unspecified if it is applied to a region of memory that was not obtained via mmap(2)." (man-pages, Release 5.05, mprotect(2)).
While I see that things should "just work" with any memory (e.g. even the example in `man mprotect` uses `memalign` to allocate the memory) I wasn't sure if there's certain distros / etc that didn't support it - I lack experiences with *Nix stuffs in general. So I just went with the naive approach of P/Invoking `mmap` directly from the C# code.

Perhaps `HAVE_SHM_OPEN_THAT_WORKS_WELL_ENOUGH_WITH_MMAP` variable means it's somewhat of an equivalent and should also work with `mprotect`?
https://github.com/dotnet/runtime/blob/c00478bb92a8eae42bf720ebd1b0bad81a34f5a4/src/libraries/Native/Unix/System.Native/pal_io.c#L311-L313

If it works and you think it's okay to use it, I'd be more than happy to use that instead. Would definitely help cleaning up this mess...

Reply:
> I'm just not sure how portable that would be

Much more portable than this manual managed implementation ;-)

Reply:
Currently I have some problems running WSL (recent Insider Preview builds broke it) so I won't be able to verify the change until I set up another VM with Linux on it, but I certainly could try that.